transaction.go

```diff
@@ -0,0 +1,489 @@
+	}
+
+	data, err := f.MarshalLedgerBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range f.Signatures {
+		sig, err := s.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		data = append(data, sig...)
+	}
+
+	return data, nil
+}
+
+// MarshalHeaderBinary marshals the transaction's header to its binary representation. See
+// UnmarshalHeaderBinary for encoding details.
+func (f *FactoidTransaction) MarshalHeaderBinary() ([]byte, error) {
+	version := varintf.Encode(f.Version)
+	data := make([]byte, TransactionHeadMinLen+len(version))
+	i := 0
+	i += copy(data[i:], version)
+
+	// Do the timestamp as 6 bytes in ms
+	ms := f.Timestamp.UnixNano() / 1e6
+	buf := bytes.NewBuffer(make([]byte, 0, 8))
+	if err := binary.Write(buf, binary.BigEndian, ms); err != nil {
+		return nil, err
+	}
+	i += copy(data[i:], buf.Bytes()[2:])
+
+	data[i] = f.InputCount
+	i += 1
+	data[i] = f.FCTOutputCount
+	i += 1
+	data[i] = f.ECOutputCount
+	i += 1
+	return data, nil
+}
+
+// MarshalBinary marshals a set of transaction ios to its binary representation. See
+// UnmarshalBinary for encoding details.
+func (ios FactoidTransactionIOs) MarshalBinary() ([]byte, error) {
+	var data []byte
+	for _, io := range ios {
+		iodata, err := io.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		data = append(data, iodata...)
+	}
+	return data, nil
+}
+
+// MarshalBinary marshals a transaction io to its binary representation. See
+// UnmarshalBinary for encoding details.
+func (io *FactoidTransactionIO) MarshalBinary() ([]byte, error) {
+	if !io.IsPopulated() {
+		return nil, fmt.Errorf("not populated")
+	}
+
+	amount := varintf.Encode(io.Amount)
+	data := make([]byte, 32+len(amount))
+	i := 0
+	i += copy(data[i:], amount)
+	i += copy(data[i:], io.Address[:])
+	return data, nil
+}
+
+// MarshalBinary marshals a transaction signature to its binary representation. See
+// UnmarshalBinary for encoding details.
+func (s *FactoidTransactionSignature) MarshalBinary() ([]byte, error) {
+	if !s.IsPopulated() {
+		return nil, fmt.Errorf("not populated")
+	}
+
+	rcdData, err := s.ReedeemCondition.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+
+	data := make([]byte, len(rcdData)+len(s.SignatureBlock))
+	i := 0
+	i += copy(data[i:], rcdData)
+	i += copy(data[i:], s.SignatureBlock)
+	return data, nil
+}
+
+const (
+	TransactionHeadMinLen = 0 + // Version length is varint
+		6 + // timestamp
+		1 + // input count
+		1 + // factoid output count
+		1 // EC output count
+
+	TransactionTotalMinLen = TransactionHeadMinLen // Coinbases have no body
+
+)
+
+// Decode will consume as many bytes as necessary to unmarshal the factoid
+// transaction. It will return the number of bytes read and an error.
+func (f *FactoidTransaction) Decode(data []byte) (i int, err error) {
+	// Because the length of an FactoidTransaction is hard to define up front, we will catch
+	// any sort of out of bound errors in a recover
+	defer func() {
```

Comment:
I do not like this. It will hide bugs and make the difficult to uncover. And doing it one place encourages it to be used again in others. 

I believe the code can be made more precise and explicit by doing better bounds checking as you go. But I'd like to hear about where it is difficult to achieve this.

Reply:
I'll attempt to add better error checking, but there was a lot of variable length things that we need to check. I will come up with the complications and work on unit tests to break them.

Reply:
the new varintf API may help some

Reply:
I removed all panic to recover defers and added enough checks to pass the tests. I don't know if some malformed data can still cause a panic, but it's much more robust now.